bat-utils/lib/runtime-kafka.js

```diff
@@ -57,25 +57,44 @@ class Kafka {
       // parallel processing on topic level
       const topicPromises = Object.keys(batchOfMessages).map(async (topic) => {
         // parallel processing on partition level
-        const partitionPromises = Object.keys(batchOfMessages[topic]).map((partition) => {
+        const topicPartitions = batchOfMessages[topic]
+        const handler = this.topicHandlers[topic]
+        const partitionKeys = Object.keys(topicPartitions)
+        for (let i = 0; i < partitionKeys.length; i += 1) {
+          const partitionKey = partitionKeys[i]
+          const partitionMessages = topicPartitions[partitionKey]
           // sequential processing on message level (to respect ORDER)
-          const messages = batchOfMessages[topic][partition]
-
-          debug('batch', topic, messages.length, messages[0])
-          return this.topicHandlers[topic](messages)
-        })
-
-        // wait until all partitions of this topic are processed and commit its offset
-        // make sure to keep batch sizes large enough, you dont want to commit too often
-        await Promise.all(partitionPromises)
-        await consumer.commitLocalOffsetsForTopic(topic)
+          for (let j = 0; j < partitionMessages.length; j += 1) {
+            const message = partitionMessages[j]
+            try {
+              await handler(message)
+              consumer.commitOffsetHard(topic, partitionKey, message.offset)
```

Comment:
does it make sense to defer the offset commit until we've either successfully processed all messages for a partition in this batch or gotten an error while processing one? that way we aren't roundtripping with kafka every time we process a single message

Reply:
yeah, i can refactor to do that